### PR TITLE
Support ONNX SAME_UPPER

### PR DESCRIPTION
**Pull request by houseroad**
1) Add support for SAME_UPPER. Looks like Caffe2 does not support SAME_LOWER by default.

2) Fix the bug in C++ ONNX C2 backend. If an attribute is renamed to empty string, we should delete it.

3) ~~Looks like we have some CuDNN fall back issue, will fix it.~~

### CODE REVIEW

File: caffe2/onnx/backend.cc [hunk 1]
```diff
@@ -304,7 +319,7 @@ Caffe2Backend::get_renamed_operators() const {
 const std::unordered_map<std::string, std::string>&
 Caffe2Backend::get_renamed_attrs() const {
   const static std::unordered_map<std::string, std::string> kRenamedAttrs{
-      {"kernel_shape", "kernels"}};
+      {"kernel_shape", "kernels"}, {"auto_pad", ""}};
```
Comment (bddppq) [inline caffe2/onnx/backend.cc]: I think you don't need this, in the `CreateConvePoolOpBase` you can remove the `auto_pad` attribute before calling `CommonOnnxNodeToCaffe2Ops`.

File: caffe2/onnx/backend.cc [hunk 2]
```diff
@@ -199,6 +200,17 @@ template <> float OnnxAttributes::get(const std::string &key) const {
   return value;
 }
 
+template <> const std::string*
+OnnxAttributes::get(const std::string &key) const {
+  const std::string* value;
```
[the remark this answers is not in the document]
Comment (houseroad) [reply caffe2/onnx/backend.cc]: Good catch

File: caffe2/onnx/backend.cc (hunk 1 above)
Comment (houseroad) [reply caffe2/onnx/backend.cc]: Sure, done :-)

File: caffe2/onnx/backend.cc [hunk 3]
```diff
@@ -244,6 +256,9 @@ OnnxAttributes::OnnxAttrToCaffe2Arg(
     const auto &attr = rewritten_onnx_attrs_.count(kv.first)
                            ? rewritten_onnx_attrs_.at(kv.first)
                            : (*kv.second);
+    if (mapper(attr.name()).empty()) {
```
Comment (bddppq) [inline caffe2/onnx/backend.cc]: this is not needed anymore

**Review by bddppq — COMMENTED** [review]
Unfortunately right now this change also needs to mirrored to the python side. @yinghai to confirm.

File: caffe2/onnx/backend.cc (hunk 3 above)
Comment (houseroad) [reply caffe2/onnx/backend.cc]: For Upsample, we still need it: https://github.com/caffe2/caffe2/blob/master/caffe2/onnx/backend.cc#L319

If we just want to remove an attribute during onnx to caffe2 conversion, probably this handling is the easiest.

File: caffe2/onnx/backend.cc [hunk 4]
```diff
@@ -199,6 +200,17 @@ template <> float OnnxAttributes::get(const std::string &key) const {
   return value;
 }
 
+template <> const std::string*
```
Comment (yinghai) [inline caffe2/onnx/backend.cc:195]: Any reason why we return a pointer instead of a copy or a const reference?

Comment (houseroad) [reply caffe2/onnx/backend.cc:195]: For the performance purpose, I avoid returning copy. 
To be consistent with TensorProto* case at line 239, use const pointer here.

File: caffe2/onnx/backend.cc [hunk 5]
```diff
@@ -411,6 +426,25 @@ Caffe2Ops Caffe2Backend::CreateConvePoolOpBase(const ModelProto &init_model,
     attr->set_i(1);
   }
 
+  if (attributes.HasAttribute("auto_pad")) {
+    auto auto_pad = attributes.get<const std::string*>("auto_pad");
```
Comment (yinghai) [inline caffe2/onnx/backend.cc:419]: `CHECK(auto_pad);`?

**Review by yinghai — APPROVED** [review]
No real blockers. Just note the comments.

**Comment by yinghai** [comment]
https://github.com/caffe2/caffe2/pull/2362 removed the python code. We can just focus on C++ code. :)

File: caffe2/onnx/backend.cc (hunk 3 above)
Comment (bddppq) [reply caffe2/onnx/backend.cc]: alright add a TODO to create another explicit list of attributes that we want to remove, like `_blacklist_caffe2_args` in the frontend

File: caffe2/onnx/backend.cc (hunk 5 above)
Comment (houseroad) [reply caffe2/onnx/backend.cc:419]: Do we have a `CHECK` macro? I didn't quite get this.

File: caffe2/onnx/backend.cc (hunk 3 above)
Comment (houseroad) [reply caffe2/onnx/backend.cc]: Sure, I will do that.

**Pull request closed without merging**